packages/query-core/src/queryClient.ts

```diff
@@ -118,6 +119,61 @@ export class QueryClient {
     return this.queryCache.find<TQueryFnData>(queryKey, filters)?.state.data
   }
 
+  ensureQueryData<
+    TQueryFnData = unknown,
+    TError = unknown,
+    TData = TQueryFnData,
+    TQueryKey extends QueryKey = QueryKey,
+  >(
+    options: WithRequired<
+      FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+      'queryKey'
+    >,
+  ): Promise<TData>
+  ensureQueryData<
+    TQueryFnData = unknown,
+    TError = unknown,
+    TData = TQueryFnData,
+    TQueryKey extends QueryKey = QueryKey,
+  >(
+    queryKey: TQueryKey,
+    options?: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+  ): Promise<TData>
+  ensureQueryData<
+    TQueryFnData = unknown,
+    TError = unknown,
+    TData = TQueryFnData,
+    TQueryKey extends QueryKey = QueryKey,
+  >(
+    queryKey: TQueryKey,
+    queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+    options?: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
```

Comment:
in the overloads with multiple parameters, we could `omit` the extra parameters from the `options` because it doesn't make sense to pass `queryKey` or `queryFn` in `options` if we have them as a separate parameter.

I did that for `useQuery`, but for nothing else. Maybe it could be a separate PR. Or maybe it doesn't matter because it will be simplified with v5 anyways 😅

Reply:
[5abe28e](https://github.com/TanStack/query/pull/4544/commits/5abe28e4c78fcf5a66aaa63dc73a6783f48cb809)
I don't think it can be changed for `arg2` and `arg3` without touching the implementation of `parseQueryArgs`. But for the second and third overload it seems to be working fine.